articles/how-to-setup-user-authentication-from-scratch-with-rails-6/index.md

```diff
@@ -0,0 +1,702 @@
+
+### Adding Controllers
+
+Routers determine which controller to use for the request,controllers will then receive the request and save or fetch data from our models.
+
+- touch `app/controllers/registrations_controller.rb`
+
+```rb
+class RegistrationsController < ApplicationController
+  # instantiates new user
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+    # stores saved user id in a session
+      session[:user_id] = @user.id
+      redirect_to root_path, notice: 'Successfully created account'
+    else
+      render :new
+    end
+  end
+
+  private
+  def user_params
+    # strong parameters
+    params.require(:user).permit(:email, :password, :password_confirmation)
+  end
+end
+```
+
+`session` stores data for one request and used in another request. This controller is responsible for creating a new user and saving it to the database.
+
+- touch `app/controllers/sessions_controller.rb`
+
+```rb
+class SessionsController < ApplicationController
+  def new; end
+
+  def create
+    user = User.find_by(email: params[:email])
+    # finds existing user, checks to see if user can be authenticated
+    if user.present? && user.authenticate(params[:password])
+    # sets up user.id sessions
+      session[:user_id] = user.id
+      redirect_to root_path, notice: 'Logged in successfully'
+    else
+      flash.now[:alert] = 'Invalid email or password'
+      render :new
+    end
+  end
+
+  def destroy
+    # deletes user session
+    session[:user_id] = nil
+    redirect_to root_path, notice: 'Logged Out'
+  end
+end
+```
+`SessionsController` provides login functionality to an existing user, also logs out a user by deleting session data.
+
+- touch `app/controllers/passwords_controller.rb`
+
+```rb
+class PasswordsController < ApplicationController
+  # allows only logged in users
+  before_action :require_user_logged_in!
+
+  def edit; end
+
+  def update
+    # update user password
+    if Current.user.update(password_params)
+      redirect_to root_path, notice: 'Password Updated'
+    else
+      render :edit
+    end
+  end
+
+  private
+  def password_params
+    params.require(:user).permit(:password, :password_confirmation)
+  end
+end
+```
+`PassWordsController` allows signed_in_user to update passwords.
+
+Update `app/controllers/application_controller.rb`
+
+```rb
+class ApplicationController < ActionController::Base
+  before_action :set_current_user
+
+  def set_current_user
+    # finds user with session data and stores it if present
+    Current.user = User.find_by(id: session[:user_id]) if session[:user_id]
+  end
+
+  def require_user_logged_in!
+    # allows only logged in user
+    redirect_to sign_in_path, alert: 'You must be signed in' if Current.user.nil?
+  end
+end
+```
+- Note inheritance hierarchy in this class, `set_current_user` will be accessed with all controllers in the subdirectory.
```

Comment:
Say something like, Since all the other controllers import the application controller, set_current_user will be accessible in all the controllers/